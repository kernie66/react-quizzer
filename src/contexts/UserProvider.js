import { createContext, useContext, useState, useEffect, useCallback } from "react";
import { useApi } from "./ApiProvider";
import { useNavigate } from "react-router-dom";
import { useQueryClient } from "@tanstack/react-query";
import { useErrorBoundary } from "react-error-boundary";
import useLoginQuery from "../hooks/useLoginQuery.js";
import { useGetQuizzerQuery } from "../hooks/useQuizzersQuery.js";

const UserContext = createContext();

export default function UserProvider({ children }) {
  // const [user, setUser] = useState();
  const [loggedInId, setLoggedInId] = useState(0);
  const api = useApi();
  const navigate = useNavigate();
  const queryClient = useQueryClient();
  const { showBoundary } = useErrorBoundary();

  const updateUserQuery = (userData) => {
    if (userData) {
      console.log("Refresh quizzers query by invalidation");
      // queryClient.setQueryData(["user", String(userData.id)], userData);
      queryClient.invalidateQueries({ queryKey: ["quizzers"] });
    }
  };

  const { isLoading, isError, data: loggedInUser } = useLoginQuery();

  const {
    // isLoading: isLoadingUser,
    data: user,
    // isError: isUserError,
    // refetch: refreshUser,
  } = useGetQuizzerQuery(loggedInId);

  useEffect(() => {
    let userId = 0;
    // Check if the user has been logged in
    if (api.isAuthenticated()) {
      // Check if the login is still valid
      try {
        if (loggedInUser) {
          userId = loggedInUser.data.userId;
          console.log("User ID:", userId);
          // let response2 = await api.get("/users/" + userId);
          // console.log("Current user:", response2.data[0]);
          // userData = response2.ok ? response2.data[0] : null;
          // updateUserQuery(userData);
        } else if (isLoading) {
          console.log("Loading login info");
        } else if (isError) {
          console.error("Error checking logged in user");
        } else {
          api.removeLogin();
        }
        // response = await api.checkLoggedIn();
      } catch (error) {
        console.error("Error checking login:", error);
        showBoundary(error);
      }
    }
    setLoggedInId(userId);
  }, [api, loggedInUser, isError, isLoading]);

  useEffect(() => {
    console.log("Current user:", user);
    updateUserQuery(user);
  }, [user]);

  const login = useCallback(
    async (username, password) => {
      const result = await api.login(username, password);
      if (result.ok) {
        const userId = result.data.id;
        const response = await api.get("/users/" + userId);
        console.log("Logged in user:", response.data[0]);
        const userData = response.ok ? response.data[0] : null;
        updateUserQuery(userData);
        // setUser(userData);
        return response;
      }
      return result;
    },
    [api],
  );

  const logout = useCallback(async () => {
    await api.logout();
    // setUser(null);
    navigate("/login");
  }, [api]);

  return <UserContext.Provider value={{ user, login, logout }}>{children}</UserContext.Provider>;
}

export function useUser() {
  return useContext(UserContext);
}
