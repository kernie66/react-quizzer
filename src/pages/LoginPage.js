import { Link, useLocation, useNavigate } from "react-router-dom";
import Body from "../components/Body";
import { useUser } from "../contexts/UserProvider";
import { useApi } from "../contexts/ApiProvider.js";
import { Trans, useTranslation } from "react-i18next";
import isValidEmail from "../helpers/isValidEmail.js";
import { trim } from "radash";
import { useErrorBoundary } from "react-error-boundary";
import { useDisclosure } from "@mantine/hooks";
import { Button, Divider, Group, Modal, PasswordInput, Text, TextInput, rem } from "@mantine/core";
import { useForm } from "@mantine/form";
import { showNotification } from "@mantine/notifications";
import { IconCheck, IconX } from "@tabler/icons-react";

export default function LoginPage() {
  const [opened, { open, close }] = useDisclosure(true);
  const { login } = useUser();
  const api = useApi();
  const { t } = useTranslation();
  const navigate = useNavigate();
  const location = useLocation();
  const { showBoundary } = useErrorBoundary();

  const form = useForm({
    initialValues: {
      username: "",
      password: "",
    },
  });

  const onSubmit = async () => {
    try {
      const username = trim(form.values.username.toLowerCase());
      const password = form.values.password;

      const errors = {};
      let userCheck = {};
      if (!username) {
        errors.username = t("username-or-email-is-missing");
      } else {
        if (isValidEmail(username)) {
          userCheck.email = username;
        } else {
          userCheck.username = username;
        }
        const existingUser = await api.get("/check", userCheck);
        if (existingUser.status !== 200) {
          errors.username = t("user-not-found");
        }
      }
      if (!password) {
        errors.password = t("password-is-missing");
      }
      if (Object.keys(errors).length === 0) {
        close();
        showNotification({
          title: t("login"),
          message: t("logging-in-username", { username }),
          color: "green",
          icon: <IconCheck style={{ width: rem(18), height: rem(18) }} />,
          autoClose: 5000,
        });
        const result = await login(username, password);
        console.log("Login result:", result);
        if (!result.ok) {
          if (result.status === 401) {
            errors.password = t("invalid-password");
          } else {
            showNotification({
              title: t("login"),
              message: t("server-error-when-logging-in"),
              color: "red",
              icon: <IconX style={{ width: rem(18), height: rem(18) }} />,
              autoClose: 5000,
            });
          }
          open();
        } else {
          let next = "/";
          if (location.state && location.state.next) {
            next = location.state.next;
          }
          navigate(next);
        }
      }
      form.setErrors(errors);
    } catch (error) {
      showBoundary(error);
    }
  };

  return (
    <Body>
      <Modal opened={opened} onClose={close} fullscreen="sm" title={t("login")} yOffset="6rem">
        <Divider mb={8} />
        <form onSubmit={form.onSubmit(onSubmit)}>
          <TextInput
            label={t("username-or-email-address")}
            {...form.getInputProps("username")}
            autoComplete="username"
            withAsterisk
            mb="md"
            data-autofocus
          />
          <PasswordInput
            label={t("password")}
            {...form.getInputProps("password")}
            autoComplete="current-password"
            withAsterisk
          />
          <Group justify="space-between" my={8} pb={8} pt={16}>
            <Button type="submit">{t("login")}</Button>
            <Button variant="outline" onClick={close}>
              {t("cancel")}
            </Button>
          </Group>
          <Divider mb={8} />
          <Text c="dimmed">
            {t("dont-have-an-account")} <Link to="/register">{t("register-here")}</Link>
          </Text>
          <Text c="dimmed">
            <Trans i18nKey="forgot-password">
              Forgot your password? Request a <Link to="/reset-request">new password here</Link>.
            </Trans>
          </Text>
        </form>
      </Modal>
    </Body>
  );
}
