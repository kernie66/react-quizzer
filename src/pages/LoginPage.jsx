import { Link, useNavigate } from "react-router-dom";
import { useUser } from "../contexts/UserProvider";
import { useApi } from "../contexts/ApiProvider";
import { Trans, useTranslation } from "react-i18next";
import isValidEmail from "../helpers/isValidEmail";
import { trim } from "radash";
import { useErrorBoundary } from "react-error-boundary";
import { useDisclosure } from "@mantine/hooks";
import { Button, Divider, Group, Modal, PasswordInput, Text, TextInput, rem } from "@mantine/core";
import { isNotEmpty, useForm } from "@mantine/form";
import { showNotification } from "@mantine/notifications";
import { TbCheck, TbX } from "react-icons/tb";
import { useState } from "react";

export default function LoginPage() {
  const [opened, { open, close }] = useDisclosure(true);
  const { login } = useUser();
  const [isLoading, setIsLoading] = useState(false);
  const api = useApi();
  const { t } = useTranslation();
  const navigate = useNavigate();
  // const location = useLocation();
  const { showBoundary } = useErrorBoundary();

  const form = useForm({
    mode: "uncontrolled",
    initialValues: {
      username: "",
      password: "",
    },
    validate: {
      username: isNotEmpty(t("username-or-email-is-missing")),
      password: isNotEmpty(t("password-is-missing")),
    },
  });

  const onSubmit = async () => {
    setIsLoading(true);
    try {
      const formValues = form.getValues();
      const username = trim(formValues.username.toLowerCase());
      form.setValues({ username: username });
      const password = formValues.password;

      const errors = {};
      const userCheck = {};
      if (!username) {
        errors.username = t("username-or-email-is-missing");
      } else {
        if (isValidEmail(username)) {
          userCheck.email = username;
        } else {
          userCheck.username = username;
        }
        try {
          const existingUser = await api.get("/check", userCheck);
          if (existingUser.status === 200) {
            console.log("existingUser", existingUser);
            console.log("User exists");
          } else {
            throw new Error("Server error", existingUser);
          }
        } catch (error) {
          if (error.response.status === 404) {
            errors.username = t("user-not-found");
          } else {
            throw new Error("Error checking username", error);
          }
        }
      }
      if (!password) {
        errors.password = t("password-is-missing");
      }
      if (Object.keys(errors).length === 0) {
        close();
        showNotification({
          title: t("login"),
          message: t("logging-in-username", { username }),
          color: "green",
          icon: <TbCheck style={{ width: rem(18), height: rem(18) }} />,
          autoClose: 5000,
        });
        const result = await login(username, password);
        console.log("Login result:", result);
        if (!result.ok) {
          if (result.status === 401) {
            errors.password = t("invalid-password");
          } else {
            showNotification({
              title: t("login"),
              message: t("server-error-when-logging-in"),
              color: "red",
              icon: <TbX style={{ width: rem(18), height: rem(18) }} />,
              autoClose: 5000,
            });
          }
          open();
        } else {
          let next = "/";
          // if (location.state && location.state.next) {
          //   next = location.state.next;
          // }
          navigate(next);
        }
      }
      form.setErrors(errors);
      setIsLoading(false);
    } catch (error) {
      setIsLoading(false);
      showBoundary(error);
    }
  };

  return (
    <>
      <Modal opened={opened} onClose={close} fullscreen="sm" title={t("login")} yOffset="6rem">
        <Divider mb={8} />
        <form onSubmit={form.onSubmit(onSubmit)}>
          <TextInput
            label={t("username-or-email-address")}
            {...form.getInputProps("username")}
            key={form.key("username")}
            autoComplete="username"
            withAsterisk
            mb="md"
            data-autofocus
          />
          <PasswordInput
            label={t("password")}
            {...form.getInputProps("password")}
            key={form.key("password")}
            autoComplete="current-password"
            withAsterisk
          />
          <Group justify="space-between" my={8} pb={8} pt={16}>
            <Button type="submit" loading={isLoading}>
              {t("login")}
            </Button>
            <Button variant="outline" onClick={close}>
              {t("cancel")}
            </Button>
          </Group>
          <Divider mb={8} />
          <Text c="dimmed">
            {t("dont-have-an-account")} <Link to="/register">{t("register-here")}</Link>
          </Text>
          <Text c="dimmed">
            <Trans i18nKey="forgot-password">
              Forgot your password? Request a <Link to="/reset-request">new password here</Link>.
            </Trans>
          </Text>
        </form>
      </Modal>
    </>
  );
}
