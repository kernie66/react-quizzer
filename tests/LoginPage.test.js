import { render, screen, fireEvent } from "@testing-library/react";
import renderer from "react-test-renderer";
import LoginPage from "../src/pages/LoginPage.js";
import UserProvider from "../src/contexts/UserProvider.js";
import FlashProvider from "../src/contexts/FlashProvider.js";
import { BrowserRouter } from "react-router-dom";
import "./i18nForTest";
import PublicRoute from "../src/components/PublicRoute.js";
import { Container } from "reactstrap";
import mockAxios from "jest-mock-axios";

afterEach(() => {
  // cleaning up the mess left behind the previous test
  mockAxios.reset();
});

describe("<LoginPage />", () => {
  it("should render login modal", () => {
    render(
      <Container>
        <BrowserRouter>
          <FlashProvider>
            <UserProvider>
              <PublicRoute>
                <LoginPage />
              </PublicRoute>
            </UserProvider>
          </FlashProvider>
        </BrowserRouter>
      </Container>,
    );
    expect(screen.getByText(/log/)).toBeInTheDocument();
    expect(mockAxios.get).toHaveBeenCalled();
  });
});

describe("<LoginPage snapshot />", () => {
  it("should show login modal", () => {
    const component = renderer.create(
      <BrowserRouter>
        <FlashProvider>
          <UserProvider>
            <PublicRoute>
              <LoginPage />
            </PublicRoute>
          </UserProvider>
        </FlashProvider>
      </BrowserRouter>,
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
