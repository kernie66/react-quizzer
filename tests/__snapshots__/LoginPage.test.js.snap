// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LoginPage snapshot /> should show login modal 1`] = `null`;
